examples/ivr/call_logic.py

```diff
@@ -0,0 +1,371 @@
+#  3. Redistributions in binary form must reproduce the above copyright notice,
+#     this list of conditions and the following disclaimer in the documentation
+#     and/or other materials provided with the distribution.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+
+import switchy
+import threading
+import re
+from functools import partial
+from collections import OrderedDict
+from switchy.marks import event_callback
+
+# Enable logging to stderr
+# Debug levels: 'INFO' for production, 'DEBUG' for development
+log = switchy.utils.log_to_stderr('INFO')
+
+
+# IVR DID pattern matching and callback registration similar to
+# `flask.route` (http://flask.pocoo.org/docs/0.10/quickstart/#routing)
+class PatternCaller(object):
+    """A `flask`-like pattern to callback registrar and invoker.
+
+    Allows for registering callback functions via decorator syntax which
+    should be called when `PatterCaller.call_matches` is invoked with a
+    matching value.
+    """
+    def __init__(self):
+        self.regex2funcs = OrderedDict()
+
+    def __call__(self, pattern, **kwargs):
+        """Decorator interface allowing you to register callback functions
+        with a regex pattern and kwargs. When `lookup` is called with a value,
+        any callable registered with a matching regex pattern will be invoked
+        with the provided kwargs.
+        """
+        def inner(func):
+            self.regex2funcs.setdefault(pattern, []).append(
+                partial(func, **kwargs))
+            return func
+
+        return inner
+
+    def call_matches(self, value, **kwargs):
+        """Perform linear lookup and callback invocation for all functions
+        registered with a matching pattern. Each function is invoked with the
+        matched value as its first argument and any arguments provided here.
+        Any kwargs which were provided at registration are also forwarded.
+        """
+        consumed = False
+        for patt, funcs in self.regex2funcs.iteritems():
+            for func in funcs:
+                match = re.match(patt, value)
+                if match:
+                    func(match=match, **kwargs)
+                    consumed = True
```

Comment:
~~~python
match = re.match(patt, value)
if match:
    consumed = True
    for func in funcs:
        func(match=match, **kwargs)
~~~